scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,165 @@
+
+The SciPy use case of the Array API is described on the following page:
+https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy
+"""
+import os
+
+import numpy as np
+# probably want to vendor it (submodule)
+import array_api_compat
+import array_api_compat.numpy
+
+__all__ = ['array_namespace', 'asarray', 'asarray_namespace']
+
+
+# SCIPY_ARRAY_API, array_api_dispatch is used by sklearn
+array_api_dispatch = os.environ.get("array_api_dispatch", False)
+SCIPY_ARRAY_API = os.environ.get("SCIPY_ARRAY_API", array_api_dispatch)
+
+_GLOBAL_CONFIG = {"SCIPY_ARRAY_API": SCIPY_ARRAY_API}
+
+
+def compliance_scipy(*arrays):
+    """Raise exceptions on known-bad subclasses.
+
+    The following subclasses are not supported and raise and error:
+    - `np.ma.MaskedArray`
+    - `numpy.matrix`
+    - Any array-like which is not Array API compatible
+    """
+    for array in arrays:
+        if isinstance(array, np.ma.MaskedArray):
+            raise TypeError("'numpy.ma.MaskedArray' are not supported")
+        elif isinstance(array, np.matrix):
+            raise TypeError("'numpy.matrix' are not supported")
+        elif not array_api_compat.is_array_api_obj(array):
+            raise TypeError("Only support Array API compatible arrays")
+
+
+def array_namespace(*arrays):
+    """Get the array API compatible namespace for the arrays xs.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays used to infer the common namespace.
+
+    Returns
+    -------
+    namespace : module
+        Common namespace.
+
+    Notes
+    -----
+    Thin wrapper around `array_api_compat.array_namespace`.
+
+    1. Check for the global switch: SCIPY_ARRAY_API. This can also be accessed
+       dynamically through ``_GLOBAL_CONFIG['SCIPY_ARRAY_API']``.
+    2. `compliance_scipy` raise exceptions on known-bad subclasses. See
+       it's definition for more details.
+
+    When the global switch is False, it defaults to the `numpy` namespace.
+    In that case, there is no compliance check. This is a convenience to
+    ease the adoption. Otherwise, arrays must comply with the new rules.
+    """
+    if not _GLOBAL_CONFIG["SCIPY_ARRAY_API"]:
+        # here we could wrap the namespace if needed
+        return np
+
+    compliance_scipy(*arrays)
+
+    return array_api_compat.array_namespace(*arrays)
+
+
+def asarray(array, dtype=None, order=None, copy=None, *, xp=None):
+    """Drop-in replacement for `np.asarray`.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    The purpose of this helper is to make it possible to share code for data
+    container validation without memory copies for both downstream use cases.
+    """
+    if xp is None:
+        xp = array_namespace(array)
+    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
```

Comment:
No one's actually doing it yet as far as I know, but this wouldn't work if someone vendors array_api_compat and tries to call a scipy function.

Reply:
Also I don't know if it makes sense to list `numpy.array_api` here. That namespace is designed to only support a strict implementation of the standard, which doesn't include `order`.

Reply:
For scikit-learn, we wanted the performance with `numpy.array_api` to be the same as `numpy`. When one explicitly sets the order, there is usually a performance reason for doing so. For example:

```python
def scipy_func(X):
    xp = array_namespace(X)
    
    # switch order for performance reasons
    X_f = asarray(X, xp, order="F")
    
    # Do some operations that require prefer F ordered.
    return xp.sum(X_f, axis=0)
```

Reply:
Actually that's wrong. I thought asarray in the spec used dlpack, but it's only numpy.asarray that does. In the spec you have to use from_dlpack (I'm not sure why they are separate).